modules/dnn/include/opencv2/dnn/dnn.hpp

```diff
@@ -71,7 +71,8 @@ CV__DNN_INLINE_NS_BEGIN
         DNN_BACKEND_HALIDE,
         DNN_BACKEND_INFERENCE_ENGINE,  //!< Intel's Inference Engine computational backend.
         DNN_BACKEND_OPENCV,
-        DNN_BACKEND_VKCOM
+        DNN_BACKEND_VKCOM,
+        DNN_BACKEND_CUDA
```

Comment:
Sorry if I already asked it before but why we have `DNN_BACKEND_CUDA` but not `DNN_BACKEND_CUDNN`?

I mean for default backend, `DNN_BACKEND_OPENCV` that means we have no extra libraries dependencies (that means it's enough to just specify `WITH_OPENCL`, in example). In case of CUDA it's enough to set `WITH_CUDA` or we need to add CUDNN dependency as well?

So the question is shall we use `DNN_BACKEND_CUDA` + `DNN_TARGET_CUDA` or `DNN_BACKEND_OPENCV` + `DNN_TARGET_CUDA`?

Reply:
Only part of layers can utilize CUDNN.
So there is expected some CUDA code (as custom layers).

These sets seems OK to me:
BACKEND_CUDA + TARGET_CUDA (current way)
or
BACKEND_CUDA + TARGET_CUDNN (may be, but there are more questions)
but
BACKEND_CUDNN + TARGET_CUDA looks strange

BACKEND_OPENCV + TARGET_CUDA - perhaphs not supported (CUDNN is required).

Reply:
> Sorry if I already asked it before but why we have DNN_BACKEND_CUDA but not DNN_BACKEND_CUDNN?

Around  25% of the layers make use of cuDNN and the rest are implemented in raw CUDA. Both `DNN_BACKEND_CUDA` and `DNN_BACKEND_CUDNN` were suggested in the #14585 issue. I choose `DNN_BACKEND_CUDA` as it was more representative (75%) than `DNN_BACKEND_CUDNN`.

> In case of CUDA it's enough to set WITH_CUDA or we need to add CUDNN dependency as well?

cuDNN is required as well.

To enable the CUDA backend:
1. set `WITH_CUDA`
2. set `WITH_CUDNN` (similar to `WITH_CUBLAS`, etc)
3. set `OPENCV_DNN_CUDA` (similar to `OPENCV_DNN_OPENCL`)

NVIDIA does not ship cuDNN along with the CUDA toolkit. The end-users have to separately install cuDNN.

The CUDA backend is a good mixture of both cuDNN and raw CUDA. The majority of the CUDA backend is in raw CUDA and it uses cuDNN's highly optimized API for some compute intensive operations such as convolutions.

I am also a bit confused about what's the most appropriate way to go about. My personal opinion is:
`DNN_BACKEND_CUDNN` : misleading as the backend is mostly in raw CUDA
`DNN_BACKEND_CUDA` : CUDA is a very general term
`DNN_BACKEND_OPENCV`: cuDNN is an additional library

Reply:
The possibility of TensorRT integration in the future might have to be considered while deciding these names.

https://docs.nvidia.com/deeplearning/sdk/tensorrt-developer-guide/index.html

Reply:
@dkurt @alalek any news on this?